### PR TITLE
Filter report data by tags

### PR DESCRIPTION
**Pull request by patrykgruszka**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ N ]
| New feature/enhancement? (use the a.x branch)      | [ Y ]
| Deprecations?                          | [ N ]
| BC breaks? (use the c.x branch)        | [ N ]
| Automated tests included?              | [ Y ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:
This PR adds possibility to filter reports with contacts by tags.

Tag filter is available in these data sources:
* Emails Sent (+ added segment filter)
* Contacts
* Multi Touch Attributions
* First Touch Attributions
* Last Touch Attributions
* Contact Point Log
* Frequency Rules
* Segment Membership
* Do Not Contact (+ added segment filter)
* Contact's segments activity log
* UTM codes (+ added segment filter)

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Add few contacts and assign tags
3. Create new Report in Reports page
4. Select Contacts (or other data source with tag filter) in "Data Source" drop down in Details tab
5. Click Add filter button to add new filter
6. Select Tags in "Column" select box and add required options
7. Save and close the report to see results

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11346"></a>

### CODE REVIEW

**Comment by npracht** [comment]
AMAZING feature 🚀

**Review by RCheesley — CHANGES_REQUESTED** [review]
Great enhancement, thanks for the PR @patrykgruszka !

One thing I noticed was that with tag filtered segments, we have four options:

- Including
- Excluding
- Empty
- Not empty

With report filters we only have two:

- Including
- Excluding

Was there a reason for this? I wonder if we should aim to be consistent where practical to allow the same filtering options.

Otherwise, working as described and a great addition! 🚀

**Comment by RCheesley** [comment]
@patrykgruszka wondered if you had any thoughts on my feedback?

**Comment by patrykgruszka** [comment]
@RCheesley yes it makes sense, I will take a closer look at this.

**Comment by patrykgruszka** [comment]
@RCheesley I've added new condition options. PR is ready to test.

**Review by RCheesley — DISMISSED** [review]
Great, thanks for adding the extra filters so that it's consistent in both places! Great feature, I hope we might see more in the reporting capabilities when it comes to tags as it's quite an important feature for the marketer!

**Review by escopecz — DISMISSED** [review]
I see no issue in the code. Thank you! 👍

File: app/bundles/ReportBundle/Tests/Event/ReportGeneratorEventTest.php [hunk 1]
```diff
@@ -285,4 +289,32 @@ public function testAddCompanyLeftJoinOnlyOnceWhenTableAlreadyJoined(): void
...
+                'column'    => 'tag',
+                'glue'      => 'and',
+                'value'     => ['3'],
+                'condition' => 'notIn',
+            ],
+        ];
+
+        $this->reportGeneratorEvent->applyTagFilter($groupExpr, $filters[0]);
+        $this->reportGeneratorEvent->applyTagFilter($groupExpr, $filters[1]);
+        Assert::assertSame('(l.id IN (SELECT DISTINCT lead_id FROM lead_tags_xref ltx WHERE ltx.tag_id IN (1, 2))) AND (l.id NOT IN (SELECT DISTINCT lead_id FROM lead_tags_xref ltx WHERE ltx.tag_id IN (3)))', $groupExpr->__toString());
```
Comment (escopecz) [inline app/bundles/ReportBundle/Tests/Event/ReportGeneratorEventTest.php]: @patrykgruszka please check the failing CI check that popped up once I merged fresh 5.x branch. For example there are missing table prefixes in the test queries.

Comment (patrykgruszka) [reply app/bundles/ReportBundle/Tests/Event/ReportGeneratorEventTest.php]: @escopecz fixed, it should be fine now.

**Review by escopecz — APPROVED** [review]

**Pull request merged**